File: leetcode/C++/kth-largest-element-in-a-stream.cpp
class KthLargest {
public:
    priority_queue<int,vector<int>,greater<int>> pq;
    int sz;
    KthLargest(int k, vector<int>& nums) {
        sort(nums.begin(), nums.end());
        for(int i=(int)nums.size()-1,j=0; i>=0&&j<k; --i,++j) pq.push(nums[i]);
        sz=k;
    }
    
    int add(int val) {
        pq.push(val); 
        if(pq.size()>sz) pq.pop();
        return pq.top();
    }
};

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */

File: leetcode/Python/kth-largest-element-in-a-stream.py
class KthLargest:

    def __init__(self, k: int, nums: List[int]):
        self.pq = list()
        self.sz = k
        nums.sort()
        j = 0
        for i in range(len(nums)-1, -1, -1):
            heapq.heappush(self.pq, nums[i])
            if j + 1 == k:
                break
            j+=1

    def add(self, val: int) -> int:
        heapq.heappush(self.pq, val)
        if len(self.pq) > self.sz:
            heapq.heappop(self.pq)
        return self.pq[0]


# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)